File: src/main/java/tech/sirwellington/alchemy/generator/GeolocationGenerators.java
package tech.sirwellington.alchemy.generator;

import tech.sirwellington.alchemy.annotations.access.NonInstantiable;

/**
 * Generators for creating Geo-Coordinates, aka longitudes and latitudes.
 */
@NonInstantiable
public final class GeolocationGenerators {

    /**
     * Object representing a latitude and longitude coordinate.
     */
    public record Coordinate(double latitude, double longitude) {
        public Coordinate {
            Checks.checkThat(latitude >= -90.0, "latitude out of bounds");
            Checks.checkThat(latitude <= 90.0, "latitude out of bounds");
            Checks.checkThat(longitude >= -180, "longitude out of bounds");
            Checks.checkThat(longitude <= 180, "longitude out of bounds");
        }
    }

    private GeolocationGenerators() throws IllegalAccessException {
        throw new IllegalAccessException("cannot directly instantiate");
    }

    /**
     * Creates valid latitudes from -90 up to and excluding 90 degrees.
     */
    public static AlchemyGenerator<Double> latitudes() {
        return NumberGenerators.doubles(-90.0, 90);
    }

    /**
     * Creates valid longitudes from -180 up to and excluding 180 degrees.
     */
    public static AlchemyGenerator<Double> longitudes() {
        return NumberGenerators.doubles(-180.0, 180.0);
    }

    /**
     * Generates valid {@link Coordinate Coordinates}.
     */
    public static AlchemyGenerator<Coordinate> coordinates() {
        var latitudes = latitudes();
        var longitudes = longitudes();
        return () -> new Coordinate(latitudes.get(), longitudes.get());
    }
}

File: src/main/java/tech/sirwellington/alchemy/generator/GeolocationGenerators.kt
/*
 * Copyright 2016 RedRoma, Inc..
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package tech.sirwellington.alchemy.generator


import tech.sirwellington.alchemy.annotations.access.NonInstantiable
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.doubles


/**
 * Generators for creating Geo-Coordinates.

 * @author SirWellington
 */
@NonInstantiable
class GeolocationGenerators
@Throws(IllegalAccessException::class)
internal constructor()
{

    init
    {
        throw IllegalAccessException("cannot instantiate")
    }

    companion object
    {

        /**
         * Creates valid latitudes from -90...90 degrees.
         * @return
         */
        fun latitudes(): AlchemyGenerator<Double>
        {
            return doubles(-90.0, 90.0)
        }

        /**
         * Creates valid longitudes from -180...180 degrees.

         * @return
         */
        fun longitudes(): AlchemyGenerator<Double>
        {
            return doubles(-180.0, 180.0)
        }
    }
}
